shared-libs/outbound/src/outbound.js

```diff
@@ -0,0 +1,216 @@
+/**
+ * Outbound is a service that creates and sends REST requests to external services based on data
+ * generated in our system, and configurations the define mappings, external service location and
+ * authentication.
+ *
+ * See: https://github.com/medic/medic-docs/blob/master/configuration/outbound.md
+ *
+ * This library deals with the conversion of the record into a REST request, and sending it to its
+ * destination. Background management of retries and so on is done in schedules, and the initial
+ * requester of an outbound push is the outbound transition.
+ */
+const objectPath = require('object-path');
+const urlJoin = require('url-join');
+const request = require('request-promise-native');
+const vm = require('vm');
+
+const secureSettings = require('@medic/settings');
+
+const OUTBOUND_REQ_TIMEOUT = 10 * 1000;
+
+// set by init()
+let logger;
+
+const fetchPassword = key => {
+  return secureSettings.getCredentials(key).then(password => {
+    if (!password) {
+      throw new Error(
+        `CouchDB config key 'medic-credentials/${key}' has not been populated. See the Outbound documentation.`
```

Comment:
Can we move this, and other, configuration validation into the transition validation function? That way it gets caught when sentinel starts rather than we a report comes in.

Reply:
This isn't config validation. Or at least, only half of it is. If you move or delete the credentials key sentinel isn't going to know about it, so we definitely have to expect this problem occurring here.

However, we can **also** check it on startup as well. I can look into that

Reply:
Wait I don't think we can do this. `transition.init` is synchronous right? It gets called from a non async function like this:

```
  if (transition.init) {
    transition.init();
  }
```

For us to check if credentials exist we need to call through to the DB.

How important do you think this is? It would be nice to do, but I'm not sure refactoring transition bootstrapping is the right thing to do right now.

Reply:
Talked offline, going to skip this for now due to refactoring and not destabiling things